src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,163 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> node;
+    private int size;
+    private Node<T> head = null;
+    private Node<T> tail = null;
+
     @Override
     public void add(T value) {
+        node = new Node<>(null, value,null);
+        if (head == null) {
+            head = tail = node;
+            size++;
+        } else {
+            tail.next = node;
+            node.prev = tail;
+            node.next = null;
+            tail = node;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        throwIndexOutOfBoundsAdd(index);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> current = findNodeByIndex(index);
+        Node<T> newNode = new Node<>(current.prev, value, current);
+        current.prev = (current.prev != null) ? current.prev.next = newNode : (head = newNode);
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            node = new Node<>(null, list.get(i), null);
+            tail.next = node;
+            node.prev = tail;
+            tail = node;
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Node<T> current = findNodeByIndex(index);
+        return current.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        throwIndexOutOfBounds(index);
+        node = new Node<>(value);
+        Node<T> current = findNodeByIndex(index);
+        if (index == 0) {
+            current.next.prev = node;
+            node.prev = null;
+            node.next = current.next;
+            head = node;
+        } else {
+            node.next = current.next;
+            node.prev = current.prev;
+            current.prev.next = node;
+            current.next.prev = node;
+        }
+        return current.item;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        throwIndexOutOfBounds(index);
+        Node<T> current = findNodeByIndex(index);
+        if (current.equals(head)) {
+            head = head.next;
+        } else {
+            current.prev.next = current.next;
+        }
+        if (current.equals(tail)) {
+            tail = tail.prev;
+        } else {
+            current.next.prev = current.prev;
+        }
+        size--;
+        return current.item;
```

Comment:
The `remove(int index)` method can be simplified by creating a separate method `unlink(Node node)`. This logic will be used in 2 methods: when we remove from the list by index and by value. So let's create a method that will take a Node that needs to be removed and unlink it. As a result, we will call it from both methods when Node is found.